src/main/java/net/minestom/server/lock/Acquirable.java

```diff
@@ -0,0 +1,121 @@
+import java.util.Objects;
+import java.util.function.Consumer;
+
+/**
+ * Represents an element which can be acquired.
+ * Used for synchronization purpose.
+ *
+ * @param <T> the acquirable object type
+ */
+public final class Acquirable<T> {
+
+    public static final ThreadLocal<Collection<Entity>> CURRENT_ENTITIES = ThreadLocal.withInitial(Collections::emptyList);
+
+    public static @NotNull Collection<@NotNull Entity> currentEntities() {
+        return CURRENT_ENTITIES.get();
+    }
+
+    @ApiStatus.Internal
+    public static void refreshEntities(@NotNull Collection<@NotNull Entity> entities) {
+        CURRENT_ENTITIES.set(entities);
+    }
+
+    private final T value;
+    private final Handler handler;
+
+    public Acquirable(@NotNull T value) {
+        this.value = value;
+        this.handler = new Handler();
+    }
+
+    /**
+     * Blocks the current thread until 'this' can be acquired,
+     * and execute {@code consumer} as a callback with the acquired object.
+     *
+     * @param consumer the acquisition consumer
+     */
+    public void acquire(@NotNull Consumer<@NotNull T> consumer) {
+        final Thread currentThread = Thread.currentThread();
+        final BatchThread elementThread = getHandler().getBatchThread();
+        Acquisition.acquire(currentThread, elementThread, () -> consumer.accept(unwrap()));
+    }
+
+    /**
+     * Executes {@code consumer} only if this element can be safely
+     * acquired without any synchronization.
+     *
+     * @param consumer the acquisition consumer
+     * @return true if the acquisition happened without synchronization,
+     * false otherwise
+     */
+    public boolean tryAcquire(@NotNull Consumer<@NotNull T> consumer) {
+        final Thread currentThread = Thread.currentThread();
+        final BatchThread elementThread = getHandler().getBatchThread();
+        if (Objects.equals(currentThread, elementThread)) {
+            consumer.accept(unwrap());
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Signals the acquisition manager to acquire 'this' at the end of the thread tick.
+     * <p>
+     * Thread-safety is guaranteed but not the order.
+     *
+     * @param consumer the consumer of the acquired object
+     */
+    public void scheduledAcquire(@NotNull Consumer<T> consumer) {
+        Acquisition.scheduledAcquireRequest(this, consumer);
+    }
+
+    /**
+     * Unwrap the contained object unsafely.
+     * <p>
+     * Should only be considered when thread-safety is not necessary (e.g. comparing positions)
+     *
+     * @return the unwraped value
+     */
+    public @NotNull T unwrap() {
+        return value;
+    }
+
+    /**
+     * Gets the {@link Handler} of this acquirable element,
+     * containing the currently linked thread.
+     *
+     * @return this element handler
+     */
+    public @NotNull Handler getHandler() {
+        return handler;
+    }
+
+    public static class Handler {
```

Comment:
Is there any reason why this couldn't be inside Acquireable? Should be documented anyway.